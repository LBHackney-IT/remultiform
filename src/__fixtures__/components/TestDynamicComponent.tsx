import PropTypes from "prop-types";
import React from "react";

import {
  DynamicComponent,
  DynamicComponentControlledProps,
} from "../../component-wrapper/DynamicComponent";

export type TestDynamicComponentProps = DynamicComponentControlledProps<
  string
> & {
  content: string;
};

export const TestDynamicComponent: React.FunctionComponent<TestDynamicComponentProps> = (
  props: TestDynamicComponentProps
): JSX.Element => {
  const { content, value, onValueChange, required, disabled } = props;

  return (
    <div>
      <div>{content}</div>
      <input
        data-testid="input"
        value={value}
        onChange={(event: React.ChangeEvent<HTMLInputElement>): void => {
          onValueChange(event.target.value);
        }}
        required={required}
        disabled={disabled}
      />
    </div>
  );
};

TestDynamicComponent.displayName = "TestDynamicComponent";

TestDynamicComponent.propTypes = {
  content: PropTypes.string.isRequired,
  ...DynamicComponent.controlledPropTypes(PropTypes.string.isRequired),
};
