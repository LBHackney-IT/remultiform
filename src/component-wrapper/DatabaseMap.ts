import PropTypes from "prop-types";

import { NamedSchema, Schema, StoreKey, StoreNames } from "../database/types";

/**
 * The options for {@link DatabaseMap}.
 */
export interface DatabaseMapOptions<
  DBSchema extends NamedSchema<string, number, Schema>,
  StoreName extends StoreNames<DBSchema["schema"]>
> {
  /**
   * The name of the {@link Database} store to fetch and update.
   */
  storeName: StoreName;

  /**
   * The key in {@link DatabaseMapOptions.storeName} to fetch and update.
   */
  key: StoreKey<DBSchema["schema"], StoreName>;
}

/**
 * A class to tell a {@link DynamicComponent} how its
 * {@link DynamicComponentControlledProps} maps to the {@link Database}.
 *
 * Unlike most generic types in this library, you will need to explicitly set
 * the type parameters on construction, as there's no way to infer a
 * {@link NamedSchema} from any of the arguments to its constructor, but they
 * constrain the possible values of its properties.
 *
 * @typeparam DBSchema - The schema of the {@link Database} this maps to.
 *
 * @typeparam StoreName - The name of the {@link Database} store this maps to.
 */
export class DatabaseMap<
  DBSchema extends NamedSchema<string, number, Schema>,
  StoreName extends StoreNames<DBSchema["schema"]>
> {
  /**
   * The proptype validator for a {@link DatabaseMap}.
   */
  static propType: PropTypes.Requireable<
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    DatabaseMap<NamedSchema<string, number, any>, string>
  > = PropTypes.exact({
    storeName: PropTypes.string.isRequired,
    key: PropTypes.oneOfType([
      PropTypes.string.isRequired,
      PropTypes.number.isRequired
    ]).isRequired
  });

  readonly storeName: StoreName;
  readonly key: StoreKey<DBSchema["schema"], StoreName>;

  constructor(options: DatabaseMapOptions<DBSchema, StoreName>) {
    const { storeName, key } = options;

    this.storeName = storeName;
    this.key = key;
  }
}
