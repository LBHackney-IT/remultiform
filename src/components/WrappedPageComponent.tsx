import { nullAsUndefined } from "null-as-undefined";
import PropTypes from "prop-types";
import React from "react";

import {
  DynamicPageComponent,
  DynamicPageComponentControlledProps,
  DynamicPageComponentType
} from "../helpers/PageComponentWrapper/DynamicPageComponent";

import { Database } from "../store/Database";
import { NamedSchema, Schema, StoreValue, StoreNames } from "../store/types";

/**
 * The proptypes for {@link WrappedPageComponent}.
 */
export interface WrappedPageComponentProps<
  Props,
  DBSchema extends NamedSchema<string, number, Schema>,
  StoreName extends StoreNames<DBSchema["schema"]>
> {
  /**
   * An open {@link Database}.
   *
   * This is usually provided by a {@link DatabaseContext.Consumer} and is
   * optional to allow waiting for {@link Database.open} to settle.
   */
  database?: Database<DBSchema> | null;

  /**
   * The {@link DynamicPageComponent} to wrap.
   */
  component: DynamicPageComponent<
    DynamicPageComponentType<Props, StoreValue<DBSchema["schema"], StoreName>>,
    Props,
    DBSchema,
    StoreName
  >;

  onChange?:
    | ((value: StoreValue<DBSchema["schema"], StoreName>) => void)
    | null;
}

interface WrappedPageComponentState<
  DBSchema extends NamedSchema<string, number, Schema>,
  StoreName extends StoreNames<DBSchema["schema"]>
> {
  value?: "" | StoreValue<DBSchema["schema"], StoreName>;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  error?: any;
  isFetching: boolean;
}

/**
 * A wrapped {@link DynamicPageComponent}.
 *
 * This handles interfacing between the
 * {@link DynamicPageComponentControlledProps} of a component and a
 * {@link Database} instance via a {@link DatabaseContext} and a
 * {@link DatabaseMap}.
 *
 * This component passes {@link DynamicPageComponentControlledProps.disabled}
 * as `true` when waiting for {@link Database} operations to settle. The
 * wrapped component should disable functionality appropriately to ensure data
 * doesn't get out of sync.
 */
export class WrappedPageComponent<
  Props,
  DBSchema extends NamedSchema<string, number, Schema>,
  StoreName extends StoreNames<DBSchema["schema"]>
> extends React.Component<
  WrappedPageComponentProps<Props, DBSchema, StoreName>,
  WrappedPageComponentState<DBSchema, StoreName>,
  never
> {
  static propTypes: PropTypes.ValidationMap<
    WrappedPageComponentProps<
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      any,
      NamedSchema<string, number, Schema>,
      StoreNames<Schema>
    >
  > = {
    database: PropTypes.instanceOf(Database),
    component: DynamicPageComponent.propType.isRequired,
    onChange: PropTypes.func
  };

  /**
   * @ignore
   */
  state: WrappedPageComponentState<DBSchema, StoreName> = {
    value: this.props.component.emptyValue,
    isFetching: false
  };

  private isUnmounted = false;

  /**
   * @ignore
   */
  async componentDidMount(): Promise<void> {
    this.isUnmounted = false;

    // We do this here rather than in `getDerivedStateFromProps` because
    // it's asynchronous, and `getDerivedStateFromProps` is expected to
    // immediately return the new state.
    await this.fetchValue();
  }

  /**
   * @ignore
   */
  async componentDidUpdate(
    prevProps: Readonly<WrappedPageComponentProps<Props, DBSchema, StoreName>>
  ): Promise<void> {
    // We only want to load from the database if it has just opened. We
    // shouldn't be changing the underlying database while this is mounted, but
    // we don't enforce that.
    if (prevProps.database !== this.props.database) {
      // We do this here rather than in `getDerivedStateFromProps` because
      // it's asynchronous, and `getDerivedStateFromProps` is expected to
      // immediately return the new state.
      await this.fetchValue();
    }
  }

  /**
   * @ignore
   */
  render(): JSX.Element {
    const { component } = this.props;
    const { error } = this.state;

    if (error) {
      console.error(error);

      // We should make this customizable.
      return <div>Something went wrong!</div>;
    }

    const { database } = this.props;
    const { value, isFetching } = this.state;
    const { key, Component, props } = component;

    const controlledProps: DynamicPageComponentControlledProps<StoreValue<
      DBSchema["schema"],
      StoreName
    >> = {
      value,
      onValueChange: this.onValueChange.bind(this),
      disabled: !database || isFetching
    };

    return <Component key={key} {...props} {...controlledProps} />;
  }

  /**
   * @ignore
   */
  componentWillUnmount(): void {
    this.isUnmounted = true;
  }

  private async fetchValue(): Promise<void> {
    if (this.isUnmounted) {
      return;
    }

    const {
      database,
      component: { databaseMap, defaultValue, emptyValue }
    } = this.props;

    if (!database || !databaseMap) {
      return;
    }

    const { storeName, key } = databaseMap;

    this.setState(state => ({ ...state, isFetching: true }));

    const stateUpdate: Partial<WrappedPageComponentState<
      DBSchema,
      StoreName
    >> = {
      isFetching: false
    };

    try {
      // If `database.get` resolves to be undefined, this will remove the
      // current `value` from the state, if one is set.
      stateUpdate.value = await database.get(storeName, key);

      if (stateUpdate.value === undefined) {
        stateUpdate.value = nullAsUndefined(defaultValue);
      }

      if (stateUpdate.value === undefined) {
        // We need to do this to keep the component as a controlled component.
        // The controlled prop must always be defined.
        stateUpdate.value = emptyValue;
      }

      // Clear the error if it was set.
      stateUpdate.error = undefined;
    } catch (err) {
      stateUpdate.error = err;
    }

    if (this.isUnmounted) {
      return;
    }

    this.setState(state => ({ ...state, ...stateUpdate }));
  }

  private onValueChange(
    value: StoreValue<DBSchema["schema"], StoreName>
  ): void {
    if (this.isUnmounted) {
      return;
    }

    const { onChange } = this.props;

    if (onChange) {
      onChange(value);
    }

    this.setState(state => ({ ...state, value }));
  }
}
