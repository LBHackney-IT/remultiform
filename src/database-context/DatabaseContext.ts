import React, { createContext } from "react";

import { Database } from "../database/Database";
import { NamedSchema, Schema } from "../database/types";

/**
 * A wrapper for a React context for passing a database connection to
 * {@link ComponentWrapper|ComponentWrappers}.
 *
 * Normally, you use this by creating an instance somewhere and sharing it
 * between components via {@link DatabaseProvider} and {@link useDatabase}.
 *
 * ```ts
 * type DBSchema = NamedSchema<
 *   "myDatabase" | "myOtherDatabase",
 *   5,
 *   {
 *     favouriteColour: {
 *       key: string;
 *       value: {
 *         r: number;
 *         g: number;
 *         b: number;
 *       };
 *     };
 *   }
 * >;
 *
 * const databasePromise = Database.open("myDatabase", 5);
 * export const DBContext = new DatabaseContext(databasePromise);
 * ```
 */
// This class exists to make it easier to enforce types in other parts of the
// library.
export class DatabaseContext<
  DBSchema extends NamedSchema<string, number, Schema>
> {
  /**
   * An open {@link Database} or a promise that will resolve to an one.
   */
  readonly database: Database<DBSchema> | Promise<Database<DBSchema>>;

  /**
   * The React context itself.
   *
   * @ignore
   */
  readonly context: React.Context<Database<DBSchema> | undefined>;

  /**
   * The consumer component for this context.
   *
   * Use it to wrap components that need the database in the usual React
   * consumer way.
   */
  readonly Consumer: React.ExoticComponent<
    React.ConsumerProps<Database<DBSchema> | undefined>
  >;

  /**
   * Create a new {@link DatabaseContext}.
   *
   * @param openDatabaseOrPromise -  An open {@link Database} or a promise that
   * will resolve to an one. Normally, you would want to pass the return value
   * of {@link Database.open} in here directly.
   */
  constructor(
    openDatabaseOrPromise: Database<DBSchema> | Promise<Database<DBSchema>>
  ) {
    this.database = openDatabaseOrPromise;
    this.context = createContext<Database<DBSchema> | undefined>(undefined);
    this.Consumer = this.context.Consumer;
  }
}
