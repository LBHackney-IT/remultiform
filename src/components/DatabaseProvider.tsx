import PropTypes from "prop-types";
import React from "react";

import { DatabaseContext } from "../helpers/DatabaseContext";

import { Database } from "../store/Database";
import { NamedSchema, Schema } from "../store/types";

/**
 * The proptypes for {@link DatabaseProvider}.
 */
export interface DatabaseProviderProps<
  DBSchema extends NamedSchema<string, number, Schema>
> {
  /**
   * An open {@link Database} or a promise that will resolve to an one.
   *
   * Normally, you would want to pass the return value of {@link Database.open}
   * in here directly.
   */
  openDatabaseOrPromise: Database<DBSchema> | Promise<Database<DBSchema>>;

  context: DatabaseContext<DBSchema>;

  /**
   * Child components of this provider.
   *
   * Anything that needs a {@link Database} to function should be contained
   * withing this subtree. Use {@link useDatabase} with
   * {@link DatabaseProviderProps.context} to access the {@link Database} from
   * within those components.
   */
  children: React.ReactNode;
}

interface DatabaseProviderState<
  DBSchema extends NamedSchema<string, number, Schema>
> {
  database?: Database<DBSchema>;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  error?: any;
}

/**
 * A React context provider for a {@link Database}.
 *
 * This will wait for the database to finish opening on mount. Until
 * {@link DatabaseProviderProps.openDatabaseOrPromise} settles, the
 * {@link DatabaseContext} will store `undefined`, the default value.
 *
 * **Important!** You will need to ensure the database closes itself if you ever
 * need to change {@link DatabaseProviderProps.openDatabaseOrPromise}, or if you
 * unmount this component. If either of those things happen, this component
 * will **throw an exception**, (unless unmounting after an error). It is
 * therefore recommended that this component is contained within an error
 * boundary that handles exceptions gracefully, probably calling
 * {@link Database.close}.
 *
 * ```js
 * const DBContext = new DatabaseContext();
 * const databasePromise = Database.open("myDatabaseWithAnError", 1);
 *
 * class Component extends React.Component {
 *   static getDerivedStateFromError(error) {
 *     return { error };
 *   }
 *
 *   state = {};
 *
 *   render() {
 *     const { error } = this.state;
 *
 *     if (error) {
 *       return <span>Something went wrong!</span>
 *     }
 *
 *     return (
 *       <DatabaseProvider
 *         context={DBContext}
 *         openDatabaseOrPromise={databasePromise}
 *       >
 *         <span>Some content that relies on the database.</span>
 *       </DatabaseProvider>
 *   }
 * }
 * ```
 */
export class DatabaseProvider<
  DBSchema extends NamedSchema<string, number, Schema>
> extends React.Component<
  DatabaseProviderProps<DBSchema>,
  DatabaseProviderState<DBSchema>,
  never
> {
  static propTypes: PropTypes.ValidationMap<
    DatabaseProviderProps<NamedSchema<string, number, Schema>>
  > = {
    openDatabaseOrPromise: PropTypes.oneOfType([
      PropTypes.instanceOf(Database).isRequired,
      PropTypes.instanceOf<
        Promise<Database<NamedSchema<string, number, Schema>>>
      >(Promise).isRequired
    ]).isRequired,

    context: PropTypes.instanceOf(DatabaseContext).isRequired,
    children: PropTypes.node.isRequired
  };

  /**
   * @ignore
   */
  state: DatabaseProviderState<DBSchema> = {};

  private isUnmounted = true;

  /**
   * @ignore
   */
  async componentDidMount(): Promise<void> {
    this.isUnmounted = false;

    const { openDatabaseOrPromise } = this.props;

    const stateUpdate: Partial<DatabaseProviderState<DBSchema>> = {};

    try {
      const database = await openDatabaseOrPromise;

      stateUpdate.database = database;
    } catch (err) {
      stateUpdate.error = err;
    }

    if (this.isUnmounted) {
      return;
    }

    this.setState(state => ({ ...state, ...stateUpdate }));
  }

  /**
   * @ignore
   */
  componentDidUpdate(
    prevProps: Readonly<DatabaseProviderProps<DBSchema>>
  ): void {
    const { error } = this.state;

    if (error) {
      throw error;
    }

    const updatedDatabase = (Object.keys(
      prevProps
    ) as (keyof typeof prevProps)[]).some(
      propName =>
        propName === "openDatabaseOrPromise" &&
        prevProps[propName] !== this.props[propName]
    );

    if (updatedDatabase) {
      this.setState({
        error: new Error(
          "Updating the openDatabaseOrPromise prop of a DatabaseProvider is unsupported"
        )
      });
    }
  }

  /**
   * @ignore
   */
  render(): JSX.Element {
    const { openDatabaseOrPromise, context, children } = this.props;
    let { database } = this.state;

    if (!database && openDatabaseOrPromise instanceof Database) {
      database = openDatabaseOrPromise;
    }

    return (
      <context.context.Provider value={database}>
        {children}
      </context.context.Provider>
    );
  }

  /**
   * @ignore
   */
  componentWillUnmount(): void {
    this.isUnmounted = true;

    const { error } = this.state;

    if (error) {
      return;
    }

    throw new Error(
      "Unmounting a DatabaseProvider outside of an error is unsupported"
    );
  }
}
