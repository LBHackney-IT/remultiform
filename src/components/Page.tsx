import PropTypes from "prop-types";
import React from "react";

import { PageComponentWrapper } from "../helpers/PageComponentWrapper/PageComponentWrapper";
import { DatabaseContext } from "../helpers/DatabaseContext";

import { NamedSchema, Schema } from "../store/types";

/**
 * The proptypes for {@link Page}.
 */
export interface PageProps<
  DBSchema extends NamedSchema<string, number, Schema>
> {
  /**
   * A context wrapper for a {@link Database} instance.
   *
   * You must provide this if any of your components are instances of
   * {@link DynamicPageComponent} for them to work as expected.
   */
  context?: DatabaseContext<DBSchema> | null;

  /**
   * An ordered array of wrapped components to display on the page.
   *
   * Create {@link PageComponentWrapper|PageComponentWrappers} with
   * {@link PageComponentWrapper.wrapStatic} or
   * {@link PageComponentWrapper.wrapDynamic}.
   */
  componentWrappers: PageComponentWrapper<DBSchema>[];
}

/**
 * A component for rendering a page of a multipage form.
 */
export const Page = <DBSchema extends NamedSchema<string, number, Schema>>(
  props: PageProps<DBSchema>
): JSX.Element => {
  const { context, componentWrappers } = props;

  if (context) {
    return (
      <context.Consumer>
        {(database): JSX.Element[] =>
          componentWrappers.map(({ render }) => render({ database }))
        }
      </context.Consumer>
    );
  }

  return <>{componentWrappers.map(({ render }) => render({}))}</>;
};

Page.propTypes = {
  context: PropTypes.instanceOf(DatabaseContext),
  componentWrappers: PropTypes.arrayOf(PageComponentWrapper.propType.isRequired)
    .isRequired
} as PropTypes.ValidationMap<PageProps<NamedSchema<string, number, Schema>>>;
