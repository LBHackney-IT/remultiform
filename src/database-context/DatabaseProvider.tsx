import PropTypes from "prop-types";
import React from "react";

import { Database } from "../database/Database";
import { NamedSchema, Schema } from "../database/types";

import { DatabaseContext } from "./DatabaseContext";

/**
 * The proptypes for {@link DatabaseProvider}.
 */
export interface DatabaseProviderProps<
  DBSchema extends NamedSchema<string, number, Schema>
> {
  context: DatabaseContext<DBSchema>;

  /**
   * Child components of this provider.
   *
   * Anything that needs a {@link Database} to function should be contained
   * withing this subtree. Use {@link useDatabase} with
   * {@link DatabaseProviderProps.context} to access the {@link Database} from
   * within those components.
   */
  children: React.ReactNode;

  /**
   * Allow unmounting outside of an error case.
   *
   * **Important!** Don't set this flag unless you know what you're doing.
   */
  allowUnmounting?: boolean | null;
}

interface DatabaseProviderState<
  DBSchema extends NamedSchema<string, number, Schema>
> {
  database?: Database<DBSchema>;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  error?: any;
}

/**
 * A React context provider for a {@link Database}.
 *
 * This will wait for the database to finish opening on mount. Until
 * {@link DatabaseContext.database} settles, the {@link DatabaseContext} will
 * store `undefined`, the default value.
 *
 * **Important!** You will need to ensure the database closes itself if you ever
 * need to change {@link DatabaseContext.database}, or if you unmount this
 * component. If either of those things happen, this component will **throw an
 * exception**, (unless unmounting after an error). It is therefore recommended
 * that this component is contained within an error boundary that handles
 * exceptions gracefully, probably calling {@link Database.close}.
 *
 * ```js
 * const databasePromise = Database.open("myDatabaseWithAnError", 1);
 * const DBContext = new DatabaseContext(databasePromise);
 *
 * class Component extends React.Component {
 *   static getDerivedStateFromError(error) {
 *     return { error };
 *   }
 *
 *   state = {};
 *
 *   render() {
 *     const { error } = this.state;
 *
 *     if (error) {
 *       return <span>Something went wrong!</span>
 *     }
 *
 *     return (
 *       <DatabaseProvider
 *         context={DBContext}
 *       >
 *         <span>Some content that relies on the database.</span>
 *       </DatabaseProvider>
 *   }
 * }
 * ```
 */
export class DatabaseProvider<
  DBSchema extends NamedSchema<string, number, Schema>
> extends React.Component<
  DatabaseProviderProps<DBSchema>,
  DatabaseProviderState<DBSchema>,
  never
> {
  static propTypes: PropTypes.ValidationMap<
    DatabaseProviderProps<NamedSchema<string, number, Schema>>
  > = {
    context: PropTypes.instanceOf(DatabaseContext).isRequired,
    children: PropTypes.node.isRequired,
  };

  /**
   * @ignore
   */
  state: DatabaseProviderState<DBSchema> = {};

  private isUnmounted = true;

  /**
   * @ignore
   */
  async componentDidMount(): Promise<void> {
    this.isUnmounted = false;

    const {
      context: { database: openDatabaseOrPromise },
    } = this.props;

    const stateUpdate: Partial<DatabaseProviderState<DBSchema>> = {};

    try {
      const database = await openDatabaseOrPromise;

      stateUpdate.database = database;
    } catch (err) {
      stateUpdate.error = err;
    }

    if (this.isUnmounted) {
      return;
    }

    this.setState((state) => ({ ...state, ...stateUpdate }));
  }

  /**
   * @ignore
   */
  componentDidUpdate(
    prevProps: Readonly<DatabaseProviderProps<DBSchema>>
  ): void {
    const { error } = this.state;

    if (error) {
      throw error;
    }

    const updatedContext = (Object.keys(
      prevProps
    ) as (keyof typeof prevProps)[]).some(
      (propName) =>
        propName === "context" && prevProps[propName] !== this.props[propName]
    );

    if (
      updatedContext &&
      this.props.context.database !== prevProps.context.database
    ) {
      this.setState({
        error: new Error(
          "Updating the context prop of a DatabaseProvider to one with a different database is unsupported"
        ),
      });
    }
  }

  /**
   * @ignore
   */
  render(): JSX.Element {
    const { context, children } = this.props;
    let { database } = this.state;
    const { database: openDatabaseOrPromise } = context;

    if (!database && openDatabaseOrPromise instanceof Database) {
      database = openDatabaseOrPromise;
    }

    return (
      <context.context.Provider value={database}>
        {children}
      </context.context.Provider>
    );
  }

  /**
   * @ignore
   */
  componentWillUnmount(): void {
    this.isUnmounted = true;

    const { allowUnmounting } = this.props;
    const { error } = this.state;

    if (error || allowUnmounting) {
      return;
    }

    throw new Error(
      "Unmounting a DatabaseProvider outside of an error is unsupported"
    );
  }
}
