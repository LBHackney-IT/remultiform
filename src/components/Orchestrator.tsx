import PropTypes from "prop-types";
import React from "react";

import {
  PageComponentWrapper,
  pageComponentWrapperPropType
} from "../helpers/PageComponentWrapper";

import { Page } from "./Page";

/**
 * A step in a multipage form.
 *
 * This represents a single step in the flow of the multipage form.
 *
 * ```ts
 * const step: Step = {
 *   key: "step-1",
 *   componentWrappers: [
 *     wrapPageComponent({
 *       key: "my-image",
 *       Component: "img",
 *       props: {
 *         src: "/path/to/my.png"
 *       }
 *     }),
 *     wrapPageComponent({
 *       key: "my-input",
 *       Component: MyInput,
 *       props: {
 *         defaultValue: "Enter something?"
 *       }
 *     })
 *   ]
 * };
 * ```
 */
export interface Step {
  /**
   * The key to uniquely identify the step
   *
   * This is referenced by {@link OrchestratorProps.currentStepKey}.
   */
  key: React.Key;

  /**
   * An ordered array of wrapped components to include in this step.
   *
   * Create {@link PageComponentWrapper|PageComponentWrappers} with
   * {@link wrapPageComponent}.
   */
  componentWrappers: PageComponentWrapper[];
}

/**
 * The proptypes for {@link Orchestrator}.
 */
export interface OrchestratorProps {
  /**
   * The key of the current step to be rendered.
   */
  currentStepKey?: React.Key | null;

  /**
   * An array of steps to possibly render based on
   * {@link OrchestratorProps.currentStepKey}.
   */
  steps: Step[];
}

/**
 * A component for orchestrating the rendering of pages for a multipage form.
 *
 * ```ts
 * const form: React.FunctionComponent = () => {
 *   return (
 *     <Orchestrator
 *       currentStepKey="step-2"
 *       steps={[
 *         {
 *           key: "step-1",
 *           componentWrappers: [
 *             wrapPageComponent({
 *               key: "my-image",
 *               Component: "img",
 *               props: {
 *                 src: "/path/to/my.png"
 *               }
 *             }),
 *             wrapPageComponent({
 *               key: "my-input",
 *               Component: MyInput,
 *               props: {
 *                 defaultValue: "Enter something?"
 *               }
 *             })
 *           ]
 *         },
 *         {
 *           key: "step-2",
 *           componentWrappers: [
 *             wrapPageComponent({
 *               key: "my-input",
 *               Component: MyInput,
 *               props: {
 *                 defaultValue: "Enter something else?"
 *               }
 *             })
 *           ]
 *         }
 *       ]}
 *     />
 *   )
 * }
 */
export const Orchestrator: React.FunctionComponent<OrchestratorProps> = (
  props: OrchestratorProps
): JSX.Element => {
  const { currentStepKey, steps } = props;

  const currentStep =
    steps.find(({ key }) => key === currentStepKey) || steps[0];

  return (
    <>
      <Page
        key={currentStep.key}
        componentWrappers={currentStep.componentWrappers}
      />
    </>
  );
};

Orchestrator.propTypes = {
  currentStepKey: PropTypes.string,
  steps: PropTypes.arrayOf(
    PropTypes.exact({
      key: PropTypes.oneOfType([
        PropTypes.string.isRequired,
        PropTypes.number.isRequired
      ]).isRequired,
      componentWrappers: PropTypes.arrayOf(pageComponentWrapperPropType)
        .isRequired
    }).isRequired
  ).isRequired
};
