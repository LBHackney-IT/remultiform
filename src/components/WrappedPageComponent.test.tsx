import React from "react";
import { ReactTestRenderer, act, create } from "react-test-renderer";

import { TestDynamicComponent } from "../__fixtures__/components/TestDynamicComponent";

import { promiseToWaitForNextTick } from "../__tests__/helpers/promise";

import { DatabaseMap } from "../helpers/PageComponentWrapper/DatabaseMap";
import { DynamicPageComponent } from "../helpers/PageComponentWrapper/DynamicPageComponent";
import { DatabaseContext } from "../helpers/DatabaseContext";

import { Database } from "../store/Database";
import { NamedSchema } from "../store/types";

import { DatabaseProvider } from "./DatabaseProvider";
import { WrappedPageComponent } from "./WrappedPageComponent";

jest.mock("../store/Database");

type TestSchema = NamedSchema<
  string,
  number,
  {
    testStore: {
      key: number;
      value: string;
    };
  }
>;

const DBContext = new DatabaseContext<TestSchema>();

const testDatabaseMap = new DatabaseMap<TestSchema, "testStore">({
  storeName: "testStore",
  key: 0
});

const testPageComponent = new DynamicPageComponent({
  key: "test-component",
  Component: TestDynamicComponent,
  props: {
    content: "test content"
  },
  defaultValue: "test default",
  databaseMap: testDatabaseMap
});

it("renders correctly with all non-context props", () => {
  const component = create(
    <WrappedPageComponent component={testPageComponent} />
  );

  expect(component).toMatchSnapshot();
});

it("renders correctly without a default value for the component", () => {
  const pageComponent = new DynamicPageComponent({
    key: "test-component",
    Component: TestDynamicComponent,
    props: {
      content: "test content"
    },
    databaseMap: testDatabaseMap
  });

  const component = create(<WrappedPageComponent component={pageComponent} />);

  expect(component).toMatchSnapshot();
});

it("renders correctly with a database context", async () => {
  const openPromise = Database.open<TestSchema>("testDBName", 1);

  const component = create(
    <DatabaseProvider openDatabaseOrPromise={openPromise} context={DBContext}>
      <DBContext.Consumer>
        {(database): JSX.Element => (
          <WrappedPageComponent
            database={database}
            component={testPageComponent}
          />
        )}
      </DBContext.Consumer>
    </DatabaseProvider>
  );

  expect(component).toMatchSnapshot();

  await openPromise;
});

it("fetches the stored value specified by the `databaseMap` when a database provided by context finishes opening", async () => {
  const databaseGetSpy = jest.spyOn(Database.prototype, "get");

  let resolveGetPromise: () => void;
  const getPromise = new Promise<void>(resolve => {
    resolveGetPromise = resolve;
  });

  databaseGetSpy.mockImplementation(async (storeName, key) => {
    await promiseToWaitForNextTick();

    resolveGetPromise();

    return `${storeName}/${key}`;
  });

  const openPromise = Database.open<TestSchema>("testDBName", 1);

  let component: ReactTestRenderer | undefined = undefined;

  await act(async () => {
    component = create(
      <DatabaseProvider openDatabaseOrPromise={openPromise} context={DBContext}>
        <DBContext.Consumer>
          {(database): JSX.Element => (
            <WrappedPageComponent
              database={database}
              component={testPageComponent}
            />
          )}
        </DBContext.Consumer>
      </DatabaseProvider>
    );

    await openPromise;
    await getPromise;
  });

  expect(databaseGetSpy).toHaveBeenCalledTimes(1);

  expect(component).toMatchSnapshot();
});

it("uses the default value when fetching the stored value returns `undefined`", async () => {
  const databaseGetSpy = jest.spyOn(Database.prototype, "get");

  let resolveGetPromise: () => void;
  const getPromise = new Promise<void>(resolve => {
    resolveGetPromise = resolve;
  });

  databaseGetSpy.mockImplementation(async () => {
    await promiseToWaitForNextTick();

    resolveGetPromise();
  });

  const openPromise = Database.open<TestSchema>("testDBName", 1);

  let component: ReactTestRenderer | undefined = undefined;

  await act(async () => {
    component = create(
      <DatabaseProvider openDatabaseOrPromise={openPromise} context={DBContext}>
        <DBContext.Consumer>
          {(database): JSX.Element => (
            <WrappedPageComponent
              database={database}
              component={testPageComponent}
            />
          )}
        </DBContext.Consumer>
      </DatabaseProvider>
    );

    await openPromise;
    await getPromise;
  });

  expect(component).toMatchSnapshot();
});

it("is disabled while the database is opening", async () => {
  const Tester = ({ disabled }: { disabled: boolean }): JSX.Element => (
    <div>{disabled ? "Is disabled" : "Is not disabled"}</div>
  );

  const pageComponent = new DynamicPageComponent({
    key: "test-component",
    Component: Tester,
    props: {},
    defaultValue: "test default",
    databaseMap: testDatabaseMap
  });

  const databaseGetSpy = jest.spyOn(Database.prototype, "get");

  databaseGetSpy.mockImplementation(async () => {});

  const openPromise = Database.open<TestSchema>("testDBName", 1);

  const component = create(
    <DatabaseProvider openDatabaseOrPromise={openPromise} context={DBContext}>
      <DBContext.Consumer>
        {(database): JSX.Element => (
          <WrappedPageComponent database={database} component={pageComponent} />
        )}
      </DBContext.Consumer>
    </DatabaseProvider>
  );

  expect(component).toMatchSnapshot();

  await openPromise;
});

it("is disabled while fetching the stored value from the database", async () => {
  const Tester = ({ disabled }: { disabled: boolean }): JSX.Element => (
    <div>{disabled ? "Is disabled" : "Is not disabled"}</div>
  );

  const pageComponent = new DynamicPageComponent({
    key: "test-component",
    Component: Tester,
    props: {},
    defaultValue: "test default",
    databaseMap: testDatabaseMap
  });

  const databaseGetSpy = jest.spyOn(Database.prototype, "get");

  let resolveGetPromise: () => void;
  const getPromise = new Promise<void>(resolve => {
    resolveGetPromise = resolve;
  });

  databaseGetSpy.mockImplementation(async () => {
    await promiseToWaitForNextTick();

    resolveGetPromise();
  });

  const database = await Database.open<TestSchema>("testDBName", 1);

  const component = create(
    <DatabaseProvider openDatabaseOrPromise={database} context={DBContext}>
      <DBContext.Consumer>
        {(database): JSX.Element => (
          <WrappedPageComponent database={database} component={pageComponent} />
        )}
      </DBContext.Consumer>
    </DatabaseProvider>
  );

  expect(component).toMatchSnapshot();

  await getPromise;
});
