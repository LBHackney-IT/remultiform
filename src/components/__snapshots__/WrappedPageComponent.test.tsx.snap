// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetches the stored value specified by the \`databaseMap\` when a database provided by context finishes opening 1`] = `
<div>
  <div>
    test content
  </div>
  <input
    data-testid="input"
    disabled={false}
    onChange={[Function]}
    value="testStore/0"
  />
</div>
`;

exports[`is disabled while fetching the stored value from the database 1`] = `
<div>
  Is disabled
</div>
`;

exports[`is disabled while the database is opening 1`] = `
<div>
  Is disabled
</div>
`;

exports[`renders correctly with a database context 1`] = `
<div>
  <div>
    test content
  </div>
  <input
    data-testid="input"
    disabled={true}
    onChange={[Function]}
  />
</div>
`;

exports[`renders correctly with all non-context props 1`] = `
<div>
  <div>
    test content
  </div>
  <input
    data-testid="input"
    disabled={true}
    onChange={[Function]}
  />
</div>
`;

exports[`renders correctly without a default value for the component 1`] = `
<div>
  <div>
    test content
  </div>
  <input
    data-testid="input"
    disabled={true}
    onChange={[Function]}
  />
</div>
`;

exports[`uses the default value when fetching the stored value returns \`undefined\` 1`] = `
<div>
  <div>
    test content
  </div>
  <input
    data-testid="input"
    disabled={false}
    onChange={[Function]}
    value="test default"
  />
</div>
`;
